tools/python/src/image.cpp

```diff
@@ -32,44 +37,32 @@ boost::python::list get_jitter_images(object img, size_t num_jitters = 1, bool d
     if (!is_rgb_python_image(img))
         throw dlib::error("Unsupported image type, must be RGB image.");
 
-    dlib::rand rnd;
-
     // Convert the image to matrix<rgb_pixel> for processing
     matrix<rgb_pixel> img_mat;
     assign_image(img_mat, numpy_rgb_image(img));
 
     // The top level list (containing 1 or more images) to return to python
     boost::python::list jitter_list;
 
+    // Size of the numpy array
+    npy_intp dims[3] = { num_rows(img_mat), num_columns(img_mat), 3};
+
     for (int i = 0; i < num_jitters; ++i) {
         // Get a jittered crop
-        matrix<rgb_pixel> crop = dlib::jitter_image(img_mat,rnd);
+        matrix<rgb_pixel> crop = dlib::jitter_image(img_mat, rnd_jitter);
         // If required disturb colors of the image
         if(disturb_colors)
-            dlib::disturb_colors(crop,rnd);
-
-        // Treat the image as a list of lists
-        boost::python::list img_as_list;
-        size_t rows = img_mat.nr();
-        size_t cols = img_mat.nc();
-        
-        for(size_t row=0; row<rows; row++) {
-            boost::python::list row_list;
-            for(size_t col=0; col<cols; col++) {
-                rgb_pixel pixel = crop(row, col);
-                boost::python::list item;
-                
-                item.append(pixel.red);
-                item.append(pixel.green);
-                item.append(pixel.blue);
-                row_list.append(item);
-            }
-            img_as_list.append(row_list);
-        }
-
-        jitter_list.append(img_as_list);
+            dlib::disturb_colors(crop, rnd_jitter);
+
+        void* img_data_copy = malloc(img_mat.nr() * width_step(crop));
+        memcpy(img_data_copy, image_data(crop), img_mat.nr() * width_step(crop));
+        PyObject *pyObj = PyArray_SimpleNewFromData(3, dims, NPY_UINT8, img_data_copy);
```

Comment:
Now you have a memory leak.

Reply:
From my understanding:
```
PyArray_ENABLEFLAGS((PyArrayObject*)pyObj, NPY_ARRAY_OWNDATA);
```
Indicated the PyArrayObject owns the data. I will reconfirm.

Reply:
Hmm, maybe that's the case.  However, it seems like you should use PyArray_SimpleNew instead.

Reply:
As per your suggestion, I am now directly allocating memory via `PyArray_SimpleNew`

The function is below:
```
boost::python::list get_jitter_images(object img, size_t num_jitters = 1, bool disturb_colors = false)
{
    if (!is_rgb_python_image(img))
        throw dlib::error("Unsupported image type, must be RGB image.");

    // Convert the image to matrix<rgb_pixel> for processing
    matrix<rgb_pixel> img_mat;
    assign_image(img_mat, numpy_rgb_image(img));

    // The top level list (containing 1 or more images) to return to python
    boost::python::list jitter_list;

    size_t rows = num_rows(img_mat);
    size_t cols = num_columns(img_mat);

    // Size of the numpy array
    npy_intp dims[3] = { num_rows(img_mat), num_columns(img_mat), 3};

    for (int i = 0; i < num_jitters; ++i) {
        // Get a jittered crop
        matrix<rgb_pixel> crop = dlib::jitter_image(img_mat, rnd_jitter);
        // If required disturb colors of the image
        if(disturb_colors)
            dlib::disturb_colors(crop, rnd_jitter);
        
        PyObject *arr = PyArray_SimpleNew(3, dims, NPY_UINT8);
        npy_uint8 *outdata = (npy_uint8 *) PyArray_DATA((PyArrayObject*) arr);

        size_t index = 0;
        for(int i=0; i<rows; i++) {
            for(int j=0; j<cols; j++) {
                rgb_pixel pixel = crop(i, j);
                outdata[index++] = pixel.red;
                outdata[index++] = pixel.green;
                outdata[index++] = pixel.blue;
            }
        }
                
        boost::python::handle<> handle(arr);
        // Append image to jittered image list
        jitter_list.append(object(handle));
    }
           
    return jitter_list;
}
```

However, do you see any problem with the previous malloc approach? `memcpy` should be faster than the `for` loop here(?)